### ISSUE TITLE
Rename Directive.services to Directive.injectables

### ISSUE THREAD

**Issue by naomiblack** [angular/angular#973]
All -- please leave a comment to vote for/against this change, or suggest a better name.

We are proposing this change to make the language more specific to what this actually does, and align better with the existing DI vocabulary.

@pkozlowski-opensource assigning to you to make the change when we have some input.

**Comment by jeffwhelpley** [comment]
How about Directive.things or Directive.stuff? j/k

Are there different types of injectables? I guess the main reason why this would not be a good idea is if there is something outside what you now consider "injectables" that would be able to be "injected" at some point. If not, then this probably makes sense and I would vote for it.

**Comment by rkirov** [comment]
There are plenty of other objects that we allow to inject, but do not need to be explicitly listed in the `services` annotation - EventEmitters, PropertySetters, parent Directives.

So `injectables` is too broad for my liking.

**Comment by TommyM** [comment]
I prefer services over injectables. Mostly because of the reasons Rado gave. Also, the services terminology has been part of angular and describes very well what they are, services!

Really liked the change from componentServices to services though!

**Comment by jwcarroll** [comment]
My knee jerk reaction is to vote against it.

It feels a little too much like pushing the framework terminology onto the developer instead of the other way around. `Injectables` is sufficiently ambiguous that I can see it leading to a really big source of confusion for a lot of people.

**Comment by Xesued** [comment]
I think I like `Directives.Injectables`.  It is a lot more clear to me. 
```javascript
@Component({
  injectables:[
    CoolService,
    TestData
  ]
})
class MyComp{
  constructor(ser:CoolService, data:TestData){/*...*/}
}
```
Looking at the code above, I don't think I would have to read the documentation to know what the injectables property is for.  I agree that `Injectables` might be broad, but unless there are plans to limit what someone can inject into their component, I think it works.

**Comment by jwcarroll** [comment]
What about `dependencies` ??? That is more inline with the domain.

**Dependency** Inversion Principle
**Dependency** Injection
Injecting **Dependencies** into a class, service, directive, etc...

I think the term `dependencies` is going to be widely recognized. Otherwise I can see myself having this conversation when teaching people about Angular 2.

Dev: "injectables... what is that?"
Me: "Those are the dependencies that get injected in. Like services and stuff."

**Comment by SekibOmazic** [comment]
+1 for `Directive.dependencies`

**Comment by TommyM** [comment]
I like dependencies as well.

**Comment by jamm** [comment]
+1 for dependencies
or injectables, if dependencies will not be accepted

**Comment by Xesued** [comment]
This might be a crazy idea, but what about removing it and creating a new annotation (if there isn't already one)?

```javascript
@Component({
  selector: 'stuff'
})
@Inject([
  MyService, SomeObject
])
class MyComponent{
  constructor(ser1:MyService, serv2:SomeObject){/*...*/}
}
```
Or `@Dependencies` :)  This way, we have a clean api of injecting into any class.

**Comment by SekibOmazic** [comment]
@Xesued or maybe:

```Javascript
@Component({
  selector: 'stuff'
})
class MyComponent{
  constructor(@Inject ser1:MyService, @Inject serv2:SomeObject){/*...*/}
}
```

**Comment by Xesued** [comment]
Oooooh.  I like that as well.  I want both please. :smiley:

**Comment by pkozlowski-opensource** [comment]
@Xesued @SekibOmazic it is not the same semantic:
* `@Inject` - this is an existing annotation and means "inject something previously declared in this injector or one of its parents"
* `services` (or injectables) means _configure_ child injector with new items that can be injected into a directive or its children

What @SekibOmazic is proposing already exists, but once again, we are looking for a construct with a different semantic here.

**Comment by Xesued** [comment]
I was unaware of the existing `@Inject`.  So, the current `services` just configures an existing Di Container (Injector?) with items to be injected?  Is the Di Container created by the `@Directive`?  If so, that makes a lot more sense why it is a property of the `@Component`.  If not, I still think and option like what I suggested would be nice (maybe named `@Dependencies`).  This might be a discussion for somewhere else.

In any case, +1 for naming it to `dependencies` or `injectables`.

**Comment by pkozlowski-opensource** [comment]
Yes, each directive will kind of have child injector, see http://victorsavkin.com/post/102965317996/angular-2-bits-unified-dependency-injection

So yes, `services` is here to configure the injector and not to express what should be injected into a directive.

**Comment by jwcarroll** [comment]
Hmm... That's a bit different than I had imagined it working. That means
it's not necessarily a set of dependencies for *that* directive... only a
way of configuring the things that *could* be injected for that component
and it's children.

Is that accurate?

If that is the case then I take back my vote for **dependencies**. Not sure
what would make it more explicit though.

**Comment by mhevery** [comment]
I don't hear a reason why we should not rename it (other than I don't like it). 

Just a remainder that `injectables` configures the injector, not what gets injected into the constructor of the component.

**Comment by pkozlowski-opensource** [comment]
OK, so reading through the comments I can extract the following proposals:
* `dependencies` - I think that this proposal was coming from misunderstanding of what `services` do currently as I think that finally this one is discarded
* `injectables` - proposal from this issue - the only "criticism" I can hear is one from @rkirov who raises 2 points: "to broad" and that there are other things that are injectable but don't have to be explicitly listed (ones coming from the element injector)
* keep `services` - haven't seen any particular arguments for keeping it and given that the term is so overloaded in ng1, we should not use it, IMO.

With the above summary it looks like `injectables` is the best proposal, although it has some issues.

I'm throwing in few ideas from me: `providers`, `dibindings`, `injectorconfig` (boy, this is long :-))

Any other proposals / objections or shell we go with `injectables`?

**Comment by mhevery** [comment]
Let's rename to `injectables` @pkozlowski-opensource  can you make it so?

**Comment by pkozlowski-opensource** [comment]
OK, `injectables` then!

**Comment by pkozlowski-opensource** [comment]
Fixed by bf7933714a884e10124ce43aa27d63791fc37e6a

**Comment by gautelo** [comment]
Even though it's closed, I'll add that I love this. I even proposed using the injectables term (`lazyInjectables([..]`) as a common name for [A1 lazy-loading of services](https://github.com/angular/angular.js/issues/11015#issuecomment-74185415) which might have been ill conceived for A1, but its much clearer than the ng-wat of services that are a kind of service that etc...
